src/Arcus.WebApi.Logging/RequestTrackingMiddleware.cs

```diff
@@ -0,0 +1,118 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Logging;
+
+namespace Arcus.WebApi.Logging
+{
+    /// <summary>
+    /// Request tracing middleware component to log every incoming HTTP request.
+    /// </summary>
+    public class RequestTrackingMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private readonly ILogger<RequestTrackingMiddleware> _logger;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="RequestTrackingMiddleware"/> class.
+        /// </summary>
+        /// <param name="next">The next pipeline function to process the HTTP context.</param>
+        /// <param name="logger">The logger to write diagnostic messages during the request tracking.</param>
+        public RequestTrackingMiddleware(
+            RequestDelegate next,
+            ILogger<RequestTrackingMiddleware> logger)
+        {
+            _next = next;
+            _logger = logger;
+        }
+
+        /// <summary>
+        /// Logs every incoming HTTP request.
+        /// </summary>
+        /// <param name="httpContext">The current HTTP context.</param>
+        public async Task Invoke(HttpContext httpContext)
+        {
+            var stopwatch = Stopwatch.StartNew();
+            if (ShouldExtractRequestBody())
+            {
+                httpContext.Request.EnableBuffering();
+            }
+
+            try
+            {
+                await _next(httpContext);
+            }
+            finally
+            {
+                stopwatch.Stop();
+                await TrackRequestAsync(httpContext, stopwatch.Elapsed);
+            }
+        }
+
+        private async Task TrackRequestAsync(HttpContext httpContext, TimeSpan duration)
+        {
+            try
+            {
+                var headers = ExtractRequestHeaders(httpContext.Request.Headers) ?? new Dictionary<string, object>();
+                var body = await ExtractRequestBodyAsync(httpContext.Request.Body) ?? new Dictionary<string, object>();
```

Comment:
I'm thinking this would be better to move to `LogRequest` as an optional parameter which is off by default, what do you think? For example, `includeRequestBody` or so?

Reply:
We could provide the same options as in the `RequestTrackingOptions` with a series of by default omitted headers and two booleans for headers/body? 🤔